### PR TITLE
Feature/prop types

### PR DESCRIPTION
**Pull request by koca**
## What is the new behavior?

Type support for all theming and component props. 
I've been playing/using this for couple of days already i can say that its a great experience. 🚀 

gif:
chakraDX

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

btw, you need to use `Volar` vscode extension to try this feature.

### CODE REVIEW

**Comment by koca** [comment]
Which syntax is better? @codebender828 @carwack 

Right now im using the first syntax which is IMO better but the second one is less code to type idk.
So what do you think both looks fine for me which one should we use?

1: 

```js
const CBoxImpl = defineComponent({
 // actual implementation
})

// export
export const CBox = CBoxImpl as ComponentWithProps<DeepPartial<BoxProps>>
```

2: 
```js
//export
export const CBox = defineComponent({
 // actual implementation
}) as ComponentWithProps<DeepPartial<BoxProps>>

```

**Comment by carwack** [comment]
> Which syntax is better? @codebender828 @carwack
> 
> Right now im using the first syntax which is IMO cleaner/better but the second one is less code to type idk.
> So what do you think both looks fine for me, which one should we use?
> 
> 1:
> 
> ```js
> const CBoxImpl = defineComponent({
>  // actual implementation
> })
> 
> // export
> export const CBox = CBoxImpl as ComponentWithProps<DeepPartial<BoxProps>>
> ```
> 
> 2:
> 
> ```js
> //export
> export const CBox = defineComponent({
>  // actual implementation
> }) as ComponentWithProps<DeepPartial<BoxProps>>
> ```

Me personally, I prefer nr 1. Because it's cleaner and easier to see what's going on.
And I do feel like it's nice to have the option to place all the exports at the end of the file 

Great job here!

File: packages/layout/src/aspect-ratio.ts [hunk 1]
```diff
@@ -68,3 +68,13 @@ export const CAspectRatio = defineComponent({
...
+
+/**
+ * Vue component used to cropping media (videos, images and maps)
+ * to a desired aspect ratio.
+ *
+ * @see Docs https://vue.chakra-ui.com/docs/layout/aspect-ratio
+ */
+export const CAspectRatio = CAspectRatioImpl as ComponentWithProps<
+  DeepPartial<AspectRatioProps>
+>
```
Comment (codebender828) [inline packages/layout/src/aspect-ratio.ts:80]: Looking great here:

I think redeclaring the component seems a bit redundant. What do you think about defining the component and then coercing the type at the end or declaring the type before the variable?

**Example 1**
```ts
export const CAspectRatio: ComponentWithProps<AspectRatioProps> = defineComponent({
  name: 'CAspectRatio',
  // ...
})
```

**Example 2**
```ts
export const CAspectRatio = defineComponent({
  name: 'CAspectRatio',
  // ...
}) as ComponentWithProps<AspectRatioProps>
```

**Review by codebender828 — CHANGES_REQUESTED** [review]
Very nice work here! Thank you @koca 

I've offered some suggestions for declaring the component types

File: packages/layout/src/aspect-ratio.ts (hunk 1 above)
Comment (koca) [reply packages/layout/src/aspect-ratio.ts:80]: > Looking great here:
> 
> I think redeclaring the component seems a bit redundant. What do you think about defining the component and then coercing the type at the end or declaring the type before the variable?
> 
> **Example 1**
> 
> ```ts
> export const CAspectRatio: ComponentWithProps<AspectRatioProps> = defineComponent({
>   name: 'CAspectRatio',
>   // ...
> })
> ```


oh i was wondering why the first syntax was not working:

image

its bec i was not using `DeepPartial`. I was like type assertion works when i use with `as` keyword but not `const CBox: TYPE = defineComponent()` 
late night coding issues :D yea lets use this syntax. Thanks :D . i'll update the pr.

**Review by codebender828 — APPROVED** [review]
Thanks for the initial work done here @koca .

Gonna merge this

**Pull request merged**